src/main/scala/calc/Compiler.scala

```diff
@@ -74,6 +78,61 @@ object Compiler {
       case Literal(value) =>
         irt.DoubleLiteral(value)
 
+      case BinaryOp(op, rhs, lhs) =>
+        val right = compileExpr(rhs, listId)
+        val left  = compileExpr(lhs, listId)
+        irt.BinaryOp(binaryOpMap(op), right, left)
+
+      case Let(name, value, body) =>
+        val irtId = irt.Ident(name.name)
+        val letVal = compileExpr(value, listId)
+        irt.Block(List(
+          irt.VarDef(irtId, letVal.tpe, false, letVal),
+          compileExpr(body, listId + (name.name -> letVal.tpe)))
+        )
+
+      case Ident(name) =>
+        listId.contains(name) match {
+          case true  => irt.VarRef(irt.Ident(name))(listId(name))
+          case false => throw new Exception(s"Identifier ${name} not defined")
+        }
+
+      case If(cond, thenp, elsep) =>
+        val ifVal = compileExpr(cond, listId)
+        irt.If(
+          irt.BinaryOp(Num_!=, ifVal, irt.DoubleLiteral(0)),
+          compileExpr(thenp, listId),
+          compileExpr(elsep, listId)
+        )(irtpe.DoubleType)
+
+      case Closure(params, body) =>
+        //Argument unboxing at the start of the function
+        val listPar = params map { x =>
+                        irt.ParamDef(irt.Ident("_" + x.name), irtpe.AnyType,
+                          false, false)
+                      }
+        val stVar = params map { x =>
+                      irt.VarDef(irt.Ident(x.name), irtpe.DoubleType, false,
+                        irt.Unbox(irt.VarRef(irt.Ident("_" + x.name))
+                          (irtpe.AnyType), 'D')
+                      )
+                    }
+        //For the capture list
+        val capList = listId.toList
```

Comment:
This is techically correct, but non efficient. You're capturing *all* values in the lexical scope of the closure.
Instead, you should capture *only* the values that are actually used in the body of the closure. This is a set known as the "free variables" for the body, if you want to Google for it.